File: tools/servidor.py
#!/usr/bin/env python3
"""
Plyscope - servidor local minimo (macOS e Linux). Equivalente ao servidor.ps1.

O navegador exige http:// (e nao file://) para carregar o motor WebAssembly
dentro de um Web Worker. Nada sai desta maquina.

Manda os mesmos cabecalhos do servidor.ps1, incluindo COOP + COEP, que ligam o
"cross-origin isolation" e liberam o SharedArrayBuffer -> Stockfish multi-thread.
Isso nao atrapalha as buscas no Chess.com e no Lichess: o COEP so barra
requisicoes "no-cors", e o fetch do app roda em modo "cors".

Manda tambem o Content-Security-Policy lido do _headers, o mesmo arquivo que
o src/build.py gera com os hashes dos <script> inline. Local e publicado rodam
sob a mesma politica de proposito: um erro de CSP tem que aparecer aqui, na
maquina de quem programa, e nao so depois do deploy.

    python3 tools/servidor.py [porta] [--sem-navegador]
"""

import functools
import http.server
import os
import sys
import threading
import webbrowser

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
PORTAS = range(8123, 8141)

MIME = {
    ".html": "text/html; charset=utf-8",
    ".js": "text/javascript; charset=utf-8",
    ".css": "text/css; charset=utf-8",
    ".wasm": "application/wasm",
    ".json": "application/json",
    ".svg": "image/svg+xml",
    ".png": "image/png",
    ".ico": "image/x-icon",
    ".pgn": "text/plain; charset=utf-8",
    ".txt": "text/plain; charset=utf-8",
    ".map": "application/json",
}


def le_csp():
    """Le o Content-Security-Policy do _headers gerado pelo src/build.py.

    Uma copia so, num arquivo so: e o que garante que o servidor local e a
    hospedagem mandem exatamente a mesma politica. Nada de repetir a string
    aqui dentro, que e como as duas acabariam divergindo.
    """
    try:
        with open(os.path.join(ROOT, "_headers"), encoding="utf-8") as f:
            for linha in f:
                if linha.strip().lower().startswith("content-security-policy:"):
                    return linha.split(":", 1)[1].strip()
    except OSError:
        pass
    return None


CSP = le_csp()


class Handler(http.server.SimpleHTTPRequestHandler):
    extensions_map = dict(http.server.SimpleHTTPRequestHandler.extensions_map, **MIME)

    def end_headers(self):
        self.send_header("Cross-Origin-Opener-Policy", "same-origin")
        self.send_header("Cross-Origin-Embedder-Policy", "require-corp")
        if CSP:
            self.send_header("Content-Security-Policy", CSP)
        self.send_header("Cache-Control", "no-cache")
        http.server.SimpleHTTPRequestHandler.end_headers(self)

    def log_message(self, *args):
        pass  # sem ruido no terminal


def main(argv):
    abrir = "--sem-navegador" not in argv and os.environ.get("PLYSCOPE_SEM_NAVEGADOR") != "1"
    pedida = [int(a) for a in argv if a.isdigit()]
    portas = pedida if pedida else list(PORTAS)

    handler = functools.partial(Handler, directory=ROOT)
    srv = porta = None
    for p in portas:
        try:
            srv = http.server.ThreadingHTTPServer(("127.0.0.1", p), handler)
            porta = p
            break
        except OSError:
            continue
    if srv is None:
        print("Nao consegui abrir uma porta local (%d-%d)." % (portas[0], portas[-1]))
        return 1

    url = "http://localhost:%d/index.html" % porta
    print("")
    print("  Plyscope rodando em %s" % url)
    print("  Deixe este terminal aberto enquanto usar o app.")
    print("  Para encerrar: Ctrl+C.")
    if not CSP:
        print("")
        print("  AVISO: nao achei o Content-Security-Policy no _headers.")
        print("  Rode 'python3 src/build.py'; sem isso o app roda aqui com regras")
        print("  mais frouxas do que no site publicado.")
    print("")
    sys.stdout.flush()
    if abrir:
        threading.Timer(0.7, lambda: webbrowser.open(url)).start()
    try:
        srv.serve_forever()
    except KeyboardInterrupt:
        print("\nEncerrado.")
    finally:
        srv.server_close()
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: tools/servidor.js
#!/usr/bin/env node
/*
 * Plyscope - servidor local minimo em Node.js.
 * Plano B do plyscope.sh para maquinas sem Python 3 (macOS sem as Command Line
 * Tools, por exemplo). Mesmos cabecalhos do servidor.ps1 e do servidor.py.
 *
 *   node tools/servidor.js [porta] [--sem-navegador]
 */
"use strict";

const http = require("http");
const fs = require("fs");
const path = require("path");
const { spawn } = require("child_process");

const ROOT = path.resolve(__dirname, "..");
const PORTAS = [];
for (let p = 8123; p <= 8140; p++) PORTAS.push(p);

const MIME = {
  ".html": "text/html; charset=utf-8",
  ".js": "text/javascript; charset=utf-8",
  ".css": "text/css; charset=utf-8",
  ".wasm": "application/wasm",
  ".json": "application/json",
  ".svg": "image/svg+xml",
  ".png": "image/png",
  ".ico": "image/x-icon",
  ".pgn": "text/plain; charset=utf-8",
  ".txt": "text/plain; charset=utf-8",
  ".map": "application/json",
};

/* Le o Content-Security-Policy do _headers gerado pelo src/build.py — a mesma
 * politica que a hospedagem manda. Uma copia so, para local e publicado nao
 * poderem divergir. */
function leCsp() {
  try {
    const txt = fs.readFileSync(path.join(ROOT, "_headers"), "utf8");
    for (const linha of txt.split("\n")) {
      const t = linha.trim();
      if (t.toLowerCase().startsWith("content-security-policy:")) {
        return t.slice(t.indexOf(":") + 1).trim();
      }
    }
  } catch (e) {}
  return null;
}
const CSP = leCsp();

const args = process.argv.slice(2);
const abrir = args.indexOf("--sem-navegador") < 0 && process.env.PLYSCOPE_SEM_NAVEGADOR !== "1";
const pedida = args.filter((a) => /^\d+$/.test(a)).map(Number);
const portas = pedida.length ? pedida : PORTAS;

const srv = http.createServer((req, res) => {
  // COOP + COEP: ligam o cross-origin isolation (SharedArrayBuffer -> multi-thread).
  res.setHeader("Cross-Origin-Opener-Policy", "same-origin");
  res.setHeader("Cross-Origin-Embedder-Policy", "require-corp");
  if (CSP) res.setHeader("Content-Security-Policy", CSP);
  res.setHeader("Cache-Control", "no-cache");

  let rel;
  try { rel = decodeURIComponent(req.url.split("?")[0].split("#")[0]); }
  catch (e) { rel = "/"; }
  rel = rel.replace(/^\/+/, "");
  if (!rel) rel = "index.html";

  const full = path.resolve(ROOT, rel);
  if (full !== ROOT && full.indexOf(ROOT + path.sep) !== 0) {
    res.writeHead(403, { "Content-Type": "text/plain; charset=utf-8" });
    res.end("403");
    return;
  }
  fs.readFile(full, (err, buf) => {
    if (err) {
      res.writeHead(404, { "Content-Type": "text/plain; charset=utf-8" });
      res.end("404 - nao encontrado: " + rel);
      return;
    }
    const tipo = MIME[path.extname(full).toLowerCase()] || "application/octet-stream";
    res.writeHead(200, { "Content-Type": tipo, "Content-Length": buf.length });
    res.end(buf);
  });
});

let i = 0;
function tentar() {
  if (i >= portas.length) {
    console.error("Nao consegui abrir uma porta local (" + portas[0] + "-" + portas[portas.length - 1] + ").");
    process.exit(1);
  }
  srv.listen(portas[i++], "127.0.0.1");
}
srv.on("error", (e) => {
  if (e.code === "EADDRINUSE" || e.code === "EACCES") tentar();
  else { console.error(String(e.message || e)); process.exit(1); }
});
srv.on("listening", () => {
  const url = "http://localhost:" + srv.address().port + "/index.html";
  console.log("");
  console.log("  Plyscope rodando em " + url);
  console.log("  Deixe este terminal aberto enquanto usar o app.");
  console.log("  Para encerrar: Ctrl+C.");
  if (!CSP) {
    console.log("");
    console.log("  AVISO: nao achei o Content-Security-Policy no _headers.");
    console.log("  Rode 'python3 src/build.py'; sem isso o app roda aqui com regras");
    console.log("  mais frouxas do que no site publicado.");
  }
  console.log("");
  if (!abrir) return;
  const cmd = process.platform === "darwin" ? "open"
    : process.platform === "win32" ? "explorer" : "xdg-open";
  try { spawn(cmd, [url], { stdio: "ignore", detached: true }).unref(); } catch (e) {}
});
tentar();
